libraries/math/src/precise_number.rs

```diff
@@ -504,6 +509,34 @@ mod tests {
         assert_eq!(root, 3); // actually 3.46572422
     }
 
+    #[test]
+    fn test_checked_mul() {
+        let number_one = PreciseNumber::new(0).unwrap();
+        let number_two = PreciseNumber::new(0).unwrap();
+        let result = number_one.checked_mul(&number_two);
+        assert_eq!(
+            result,
+            Option::Some(PreciseNumber {
+                value: U256::from(0)
+            })
+        );
+
+        let number_one = PreciseNumber::new(2).unwrap();
+        let number_two = PreciseNumber::new(2).unwrap();
+        let result = number_one.checked_mul(&number_two).unwrap();
+        assert_eq!(result, PreciseNumber::new(4).unwrap());
+
+        let number_one = PreciseNumber::new(340282366920938463463374607431768211455).unwrap();
+        let number_two = PreciseNumber::new(340282366920938463463374607431768211455).unwrap();
+        let result = number_one.checked_mul(&number_two);
+        assert_eq!(result, Option::None);
+
+        let number_one = PreciseNumber::new(340282366920938463463374607431768211454).unwrap();
+        let number_two = PreciseNumber::new(340282366920938463463374607431768211455).unwrap();
+        let result = number_one.checked_mul(&number_two);
+        assert_eq!(result, Option::None);
```

Comment:
How about a test right at the max possible, and one that just goes right over?

Reply:
You mean testing with #[allow(arithmetic_overflow)] and #[should_panic] ?

Reply:
Not exactly -- the two tests at `u128::MAX` and `u128::MAX - 1` mostly give the same result for the multiplication, so the more interesting result would be a multiplication that yields an inner representation of `U256::MAX` and one that yields `U256::MAX + 1`, which would become `None`. You could do:
```
let number_one = PreciseNumber { value: U256::from(u128::MAX) };
let number_two = PreciseNumber { value: U256::from(u128::MAX) };
let result = number_one.checked_mul(&number_two);
assert_eq!(result.value, U256::MAX);
```
and then
```
let number_one = PreciseNumber { value: U256::from(u128::MAX) + 1 };
let number_two = PreciseNumber { value: U256::from(u128::MAX) };
let result = number_one.checked_mul(&number_two);
assert_eq!(result, None);
```

Reply:
@joncinque 

-1 is for testing this
```
                let value = if self.value >= rhs.value {
                    self.value.checked_div(one())?.checked_mul(rhs.value)?
                } else {
                    rhs.value.checked_div(one())?.checked_mul(self.value)?
                };
```

+1 doesn't become None from what I see here.

Can you merge and do some experiments on your own? Maybe something doesn't work as expected. It's the first project where I'm dealing with U256 (I know, it's lame excuse :)).

Reply:
And BTW. I'm not sure if

```
Some(Self { value })
```
at line 164 is not a dead code.

Reply:
I'll push a test for this onto this branch so you can see what I mean.